### ISSUE TITLE
Save/Load for Collections

### ISSUE THREAD

**Issue by omidb** [mpollmeier/gremlin-scala#73]
Hi,

I saw save/load feature. Why some collections like Array are not in `persistableTypes`. Maybe @joan38 can also answer this.

**Comment by joan38** [comment]
Hi omidb,

https://github.com/tinkerpop/rexster/wiki/Property-Data-Types
We can apparently persist Lists and Maps of primitive types in Gremlin.
I will do a PR with this feature.

Cheers

**Comment by omidb** [comment]
Thanks, what do you think abou nested case classes ? We can create Map for them

Sent from my Windows Phone
________________________________

**Comment by joan38** [comment]
What is the reason for that?
It sound a bit dodgy to me in case you need to refactor/extract your class out.

**Comment by omidb** [comment]
So most of the time data is a little bit more complex that a simple case classes and you don't want to put it as another node:
```scala
case class Person(id:String, name:String, address: Address)
case class Address(street:String, ....)
```
So we can do this with nested Maps, we can write  a macro to convert any case classes to a Map,(even shapeless may have this), does it sound right? (I am thinking about something like Bson in ReactiveMongo, or Json implementation in Play)

**Comment by joan38** [comment]
That doesn't sound like a bad idea.
I'll try an implementation.

See that we are now supporting the collections.

**Comment by omidb** [comment]
@joan38 thanks for collection
Here https://github.com/themillhousegroup/sausagefactory is Reflection way to it
I was looking for macro based implementation, here is my question on Stackoverflow: http://stackoverflow.com/questions/31528379/scala-macros-for-nested-case-classes-to-map-and-other-way-around?noredirect=1#comment51023005_31528379 would maybe helps

**Comment by joan38** [comment]
If you can have a way to generate a `toMap()` / `fromMap()` on case classes with macros we would be able to avoid reflection even on the save / load :)

**Comment by omidb** [comment]
I'm trying to do that, there exist one for JSON, in Play Json library you write an implicit conversion like this and it will do it:
```scala
implicit val myClaseClassFormat = Json.format[MyCaseClass]
```
and by providing the implicit in the context of JsonWriter it will use it to generate Json. It's a very neat implementation. 
I can write a macro to convert the case class to Map[String,Any] but it's not working for nested case classes.

**Comment by joan38** [comment]
Can you show me how you write the macro? In a branch or a Gist?
I was thinking that we actually don't even need an annotation but the save() method should be a def macro.

**Comment by joan38** [comment]
@omidb could you tell me where do you persist Collections?
The reason why I am asking is because Titan doesn't seems to support any collections:
http://s3.thinkaurelius.com/docs/titan/0.9.0-M2/schema.html#d0e985
And I can't find any info for Neo4J.

It works with TinkerGraph but this one actually accept any types.

Thanks

**Comment by omidb** [comment]
@joan38 collections in the case classes? Doesn't Rexter deal with List and Map? currently we can dal with a List[T] as long as T is primitive type. But I'm trying to add the support for List[T] where T is a case class:
```scala
case class Person(name:String, addresses: List[Address])
case class Address(zip:Int)

val people = Person("joan", List(Address(1111),Address(2222)))
//it gets converted to:
//Map("name" -> "joan", "addresses" -> List(Map("zip" -> 1111),Map("zip" -> 2222)))
I think if we add these and tuples we will have very general support of case classes
```

**Comment by joan38** [comment]
Rexter seems to support the List and Map.
Even better TinkerGraph does support any objects.
But I can't find any concret production DB that support List and Map.
In Titan we can actually persist any object but we need to fulfil some requirements:
http://s3.thinkaurelius.com/docs/titan/0.9.0-M2/serializer.html#d0e10007

I'm worried that we are actually doing something that is not actually useful as it either supports all objects (TinkerGraph) or no objects (unless custom serialisation).

**Comment by omidb** [comment]
mmhh, that's weird, if there is some capability that just dump the object into a vertex, why don't we have any function for it in Tinkerpop, or do we?

**Comment by joan38** [comment]
Tinkerpop seems to take any object. As I have been able to assess I can persist any object in TinkerGraph.
It seems that for Titan it's not that easy, there is an interface at least to inherit from which is not a Tinkerpop interface but a Titan interface.

**Comment by omidb** [comment]
Can you test the code in Titan, I also will try the same thing in Neo4J. TinkerGraph is in memory database but it seems anything that works there should also work with other things

**Comment by omidb** [comment]
I created #80 to talk about other issue, but this issue has been solved in #77
